src/EA.Iws.RequestHandlers/NotificationMovements/BulkReceiptRecovery/ReceiptRecoveryShipmentMustBePrenotifiedRule.cs

```diff
@@ -0,0 +1,53 @@
+﻿namespace EA.Iws.RequestHandlers.NotificationMovements.BulkReceiptRecovery
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Core.Movement;
+    using Core.Movement.BulkReceiptRecovery;
+    using Core.Rules;
+    using Domain.Movement;
+
+    public class ReceiptRecoveryShipmentMustBePrenotifiedRule : IReceiptRecoveryContentRule
+    {
+        private readonly IMovementRepository repo;
+
+        public ReceiptRecoveryShipmentMustBePrenotifiedRule(IMovementRepository repo)
+        {
+            this.repo = repo;
+        }
+
+        public async Task<ReceiptRecoveryContentRuleResult<ReceiptRecoveryContentRules>> GetResult(List<ReceiptRecoveryMovement> movements, Guid notificationId)
+        {
+            var actualMovements = await repo.GetAllMovements(notificationId);
+
+            List<int> shipments = new List<int>();
+            MessageLevel result = MessageLevel.Success;
+
+            foreach (var movement in movements.Where(p => !p.MissingReceivedDate && !p.MissingRecoveredDisposedDate))
```

Comment:
Please use `p.ReceivedDate.HasValue` for checking the date value. Same applies for the rececovered/disposed date property. The Missing.... `bool` properties indicates whether the cell in the spreadsheet is blank. If it's not, then this is then parsed into their correct types - in this case it's DateTime. Now the user may provide some random value that can't be parsed into a DateTime, which will mean the property will be set to `null`.

Reply:
Actually, ignore my first comment on this. This is correct using the Missing... properties.